utils/pubsub/pubsub_test.go

```diff
@@ -9,116 +9,81 @@ func TestPubSub_two_consumers_slow_and_fast(t *testing.T) {
+	normal := bus.Subscribe(ctx, topic)
 
 	bus.Publish(topic, 1)
-	t.Log("slow consumer started")
-	normal.startConsuming(ctx)
-	t.Log("normal consumer started")
-
-	// normal consumer should consume 1, 2, 3, 4 & 5
-	v := <-normal.out
-	if v != 1 {
-		t.Errorf("Expected normal consumer to have consumed 1, instead got %d.", v)
-	}
-
-	bus.Publish(topic, 2)
-	v = <-normal.out
-	if v != 2 {
-		t.Errorf("Expected normal consumer to have consumed 2, instead got %d.", v)
-	}
-
-	bus.Publish(topic, 3)
-	v = <-normal.out
-	if v != 3 {
-		t.Errorf("Expected normal consumer to have consumed 3, instead got %d.", v)
+	v := <-normal
+	if v.Data.(int) != 1 {
+		t.Errorf("Expected slow consumer to have consumed 5, instead got %d.", v.Data.(int))
 	}
-
-	bus.Publish(topic, 4)
-	v = <-normal.out
-	if v != 4 {
-		t.Errorf("Expected normal consumer to have consumed 4, instead got %d.", v)
+	v = <-slow
+	if v.Data.(int) != 1 {
+		t.Errorf("Expected slow consumer to have consumed 5, instead got %d.", v.Data.(int))
 	}
 
-	bus.Publish(topic, 5)
-	v = <-normal.out
-	if v != 5 {
-		t.Errorf("Expected normal consumer to have consumed 5, instead got %d.", v)
+	t.Log("normal consumer gets ahead by 4 publishes")
+	numbers := []int{2, 3, 4, 5}
+	for _, n := range numbers {
+		bus.Publish(topic, n)
+		// normal consumer should consume 1, 2, 3, 4 & 5
+		v := <-normal
+		if v.Data.(int) != n {
+			t.Errorf("Expected normal consumer to have consumed 1, instead got %d.", v.Data.(int))
+		}
 	}
 
-	slow.startConsuming(ctx)
-	// slow consumer should consume 1 & 5
-	v = <-slow.out
-	if v != 1 {
-		t.Errorf("Expected slow consumer to have consumed 1, instead got %d.", v)
-	}
-	v = <-slow.out
-	if v != 5 {
-		t.Errorf("Expected slow consumer to have consumed 5, instead got %d.", v)
+	// slow consumer
+	t.Log("slow consumer should consume only the latest one, 5")
+	v = <-slow
+	if v.Data.(int) != 5 {
+		t.Errorf("Expected slow consumer to have consumed 5, instead got %d.", v.Data.(int))
 	}
-
 }
 
 func TestPubSub_late_subscription_gets_latest_value(t *testing.T) {
 	const topic = "topic"
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	bus := NewPublishSubscriber(ctx)
+	bus := &PublishSubscriber{}
 
 	// Given I publish to a topic
 	bus.Publish(topic, 1)
 	bus.Publish(topic, 2)
 
 	// When I subscribe to the topic after the published event
-	normal := testConsumer{}
-	normal.subscribe("normal", topic, bus)
-	normal.startConsuming(ctx)
+	normal := bus.Subscribe(ctx, topic)
 
 	// Then the new subscriber receives the latest published value
-	v := <-normal.out
-	if v != 2 {
-		t.Errorf("Expected late consumer to have consumed latest value 2, instead got %d.", v)
+	v := <-normal
+	if v.Data.(int) != 2 {
+		t.Errorf("Expected late consumer to have consumed latest value 2, instead got %d.", v.Data.(int))
 	}
 
 }
 
-type testConsumer struct {
-	name string
-	in   chan DataEvent
-	out  chan int
-}
+func TestPubSub_Close(t *testing.T) {
```

Comment:
It would be useful to add another test where we are cancelling the subscription's context